File: clsDBHelper.vb
Option Strict Off
Imports System.Data.SqlClient
Public Class clsDBHelper

    Public _SQLConn As SqlConnection

    Public DataTable As New DataTable

    Public Sub New(ByRef AccessConn As SqlConnection)

        _SQLConn = AccessConn

    End Sub

    Public Function ExecuteStoredProc(ByVal strStoredProc As String, ByVal _parameters() As SqlParameter) As Boolean


        Try

            Dim cm As SqlCommand = New SqlCommand(strStoredProc, _SQLConn)
            cm.CommandType = CommandType.StoredProcedure

            If (Not (_parameters) Is Nothing) Then
                For Each p As SqlParameter In _parameters
                    cm.Parameters.Add(p)

                Next
            End If

            cm.ExecuteNonQuery()


            Return True

        Catch ex As Exception

            Throw New ApplicationException(ex.Message)

            Return False

        End Try
    End Function


    Public Function TableExist(ByVal tblName As String) As Boolean

        '  Specify restriction to get table definition schema
        '  For reference on GetSchema see:
        '  http://msdn2.microsoft.com/en-us/library/ms254934(VS.80).aspx
        Dim restrictions As String() = New String(2) {}
        restrictions(2) = tblName
        Dim dbTbl As DataTable = _SQLConn.GetSchema("Tables", restrictions)

        If (dbTbl.Rows.Count = 0) Then
            ' Table does not exist
            Return False
        Else
            ' Table exists
            Return True
        End If

        dbTbl.Dispose()

    End Function


    ''' <summary>
    ''' Returns a complete dataset from ; delimited queries
    ''' </summary>
    ''' <param name="strQuery"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetDataSet(ByVal strQuery As String) As System.Data.DataSet

        'NOTE:  Since the JET engine does not support multiple queries, we need to create separate data adapters and add them to the dataset..

        'trim the last ; off if it exists...
        strQuery = strQuery.Trim()
        If (strQuery.Substring(strQuery.Length - 1, 1) = ";") Then strQuery = Left(strQuery, Len(strQuery) - 1)


        Dim strQueries() As String = Split(strQuery, ";")
        Dim da(strQueries.Length - 1) As SqlDataAdapter

        Dim ds As New System.Data.DataSet

        For x As Int32 = 0 To strQueries.Length - 1

            '2. Create the command object, passing in the SQL string    
            Dim cmd As New SqlCommand(strQueries(x), _SQLConn)

            '3. Create the DataAdapter
            da(x) = New SqlDataAdapter()
            da(x).SelectCommand = cmd

            '4. Populate the DataSet and close the connection

            da(x).Fill(ds, "Table" & x.ToString())

        Next

        'Return the DataSet
        Return ds


    End Function



    Public Function GetRecordsetDT(ByVal strQuery$) As DataTable
        'Function fills public "DataTable" with results of recordset.  Returns false on error.

        Dim da As SqlDataAdapter

        Dim dt As New DataTable


        da = New SQLDataAdapter(strQuery, _SQLConn)
        da.Fill(dt)

        Return dt





    End Function


    Public Function GetRecordset(ByVal strQuery$) As Boolean
        'Function fills public "DataTable" with results of recordset.  Returns false on error.

        Dim da As SQLDataAdapter
        GetRecordset = False
        If Not DataTable Is Nothing Then DataTable.Clear()


        da = New SQLDataAdapter(strQuery, _SQLConn)
        da.Fill(DataTable)

        GetRecordset = True
        da = Nothing


    End Function

    'Public Sub FillGrid(ByRef grid As ComponentFactory.Krypton.Toolkit.KryptonDataGridView, ByVal ParamArray strFieldList() As Object)
    '    Dim x As Integer, i As Integer
    '    Dim objCells() As Object

    '    grid.Rows.Clear()
    '    With DataTable

    '        ReDim objCells(UBound(strFieldList))
    '        For x = 0 To .Rows.Count - 1

    '            For i = 0 To UBound(strFieldList)

    '                If strFieldList(i) <> "" Then

    '                    objCells(i) = IIf(IsDBNull(.Rows(x).Item(strFieldList(i))), "", .Rows(x).Item(strFieldList(i)))

    '                End If
    '            Next i

    '            grid.Rows.Add(objCells)

    '        Next x

    '    End With




    'End Sub

    Public Function GetLastIdentity() As Integer

        Dim objCommand As SQLCommand
        GetLastIdentity = -1



        objCommand = _SQLConn.CreateCommand()
        objCommand.CommandText = "SELECT @@IDENTITY AS NewID;"

        GetLastIdentity = CInt(objCommand.ExecuteScalar())

        objCommand = Nothing




    End Function


    Public Function ExecuteScalar(ByVal strQuery$) As Object
        'Use ExecuteScalar method to return a single value (for example, an aggregate value) 
        'from the database. This requires less code than using the ExecuteReader method, and then performing 
        'the operations necessary to generate the single value using the data returned by a SqlDataReader.

        'The function returns field object
        Dim objCommand As SQLCommand
        ExecuteScalar = Nothing


        objCommand = _SQLConn.CreateCommand()
        objCommand.CommandText = strQuery

        ExecuteScalar = objCommand.ExecuteScalar()

        objCommand = Nothing




    End Function

    Public Function FixForApostrophe(ByVal strFieldContents As String) As String
        ' replace each apostrophe with two apostrophes

        If Trim(strFieldContents) = "" Then Return ""

        FixForApostrophe = Replace(strFieldContents, Chr(39), "''")

    End Function


    Public Function BuildUpdateQry(ByVal strTableName As String, ByVal ParamArray objFieldsAndValues() As Object) As String
        'Function returns a formatted insert query only. Makes NO connection to the database. 
        Dim qry As String
        Dim x As Int16
        qry = ("Update " _
                    + (strTableName + " Set "))
        x = 0
        Do While (x <= (objFieldsAndValues.GetLength(0) - 1))

            If (objFieldsAndValues((x + 1)) Is Nothing) Then
                objFieldsAndValues((x + 1)) = ""
            End If
            Dim str As String = objFieldsAndValues((x + 1)).GetType.Name
            str = str.ToUpper
            Select Case (str)
                Case "BOOLEAN"
                    If (CType(objFieldsAndValues((x + 1)), Boolean) = True) Then
                        qry = (qry _
                                    + (objFieldsAndValues(x) + "= 1, "))
                    Else
                        qry = (qry _
                                    + (objFieldsAndValues(x) + "= 0, "))
                    End If
                Case "INT32", "INT16", "INT64", "DOUBLE", "SINGLE"
                    qry = (qry _
                                + (objFieldsAndValues(x) & ("=" _
                                + (objFieldsAndValues((x + 1).ToString()) & ", "))))
                Case "STRING", "DATETIME"
                    If (objFieldsAndValues((x + 1)).ToString = "NULL") Then
                        qry = (qry _
                                    + (objFieldsAndValues(x) + "= NULL, "))
                    Else
                        qry = (qry _
                                    + (objFieldsAndValues(x) + ("= '" _
                                    + (FixForApostrophe(objFieldsAndValues((x + 1)).ToString) + "', "))))
                    End If

                Case "GUID"

                    If (objFieldsAndValues((x + 1)).ToString = "NULL") Then
                        qry = (qry _
                                    + (objFieldsAndValues(x) + "= NULL, "))
                    Else
                        qry = (qry _
                                    + (objFieldsAndValues(x) + ("= {" _
                                    + (FixForApostrophe(objFieldsAndValues((x + 1)).ToString) + "}, "))))
                    End If


                Case Else
                    qry = (qry _
                                + (objFieldsAndValues(x) + ("=" _
                                + (objFieldsAndValues((x + 1)) + ", "))))
            End Select
            x = (x + 2)
        Loop
        Return qry.Substring(0, (qry.Length - 2))
    End Function



    Public Function BuildInsertQry(ByVal strTableName As String, ByVal ParamArray strFieldsAndValues() As Object) As String
        'Function returns a formatted insert query only.  Makes NO connection to the database.

        Dim x As Int16
        BuildInsertQry = "Insert Into " & strTableName & " ("
        For x = 0 To UBound(strFieldsAndValues) Step 2

            BuildInsertQry += strFieldsAndValues(x) & ","

        Next
        BuildInsertQry = Left(BuildInsertQry, Len(BuildInsertQry) - 1) & ") VALUES ("

        For x = 1 To UBound(strFieldsAndValues) Step 2

            Select Case UCase(strFieldsAndValues(x).GetType.Name)

                Case "BOOLEAN" : BuildInsertQry += IIf(strFieldsAndValues(x), 1, 0) & ","
                Case "INT64", "INT32", "INT16" : BuildInsertQry += strFieldsAndValues(x) & ","
                Case "STRING", "DATETIME" : BuildInsertQry += "'" & FixForApostrophe(strFieldsAndValues(x)) & "',"
                Case "GUID" : BuildInsertQry &= "{" & strFieldsAndValues(x).ToString() & "},"
                Case "DBNULL" : BuildInsertQry &= "NULL, "
                Case Else : BuildInsertQry += strFieldsAndValues(x) & ","

            End Select

        Next
        BuildInsertQry = Left(BuildInsertQry, Len(BuildInsertQry) - 1) & ")"


    End Function

    Public Function ExecuteNonQuery(ByVal strQuery$) As Boolean
        'No error trapping here.  We want errors to bubble through to the calling function.

        Dim objCommand As SqlCommand


        objCommand = New SqlCommand(strQuery, _SQLConn)
        objCommand.ExecuteNonQuery()
        ExecuteNonQuery = True


        objCommand = Nothing





    End Function


    Protected Overrides Sub Finalize()

        Try

            DataTable = Nothing
          

        Catch ex As Exception
        End Try


        MyBase.Finalize()
    End Sub
End Class

File: clsDBHelper.cs
using System;
using System.Collections.Generic;
using System.Text;
using Microsoft.VisualBasic;
using System.Data.SqlClient;
using System.Data;
using System.Windows.Forms;

namespace MachineStatus
{

    public class clsDBHelper
    {
        
        public clsDBHelper(SqlConnection conn)
        {
            _conn = conn;
        }
        
        private SqlConnection _conn;

        public System.Data.DataTable DataTbl = new System.Data.DataTable();


        /// <summary>
        /// Executes a stored proc and returns a data table with the results.
        /// </summary>
        /// <param name="strStoredProc"></param>
        /// <param name="_parameters"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public DataTable ExecuteStoredProcDT(string strStoredProc, SqlParameter[] _parameters)
        {


            try
            {
                
                SqlCommand cm = new SqlCommand(strStoredProc, _conn);
                cm.CommandType = CommandType.StoredProcedure;

                // Check to see if there are any paramters being passed in.
                // If there are none, then do not add and execute the SP
                if ((((_parameters) != null)))
                {
                    foreach (SqlParameter p in _parameters)
                    {
                        cm.Parameters.Add(p);
                    }
                }

                SqlDataAdapter da = new SqlDataAdapter(cm);
                DataSet ds = new DataSet();
                DataTable dt = new DataTable();

                da.Fill(ds, "rs");
                dt = ds.Tables["rs"];
                return dt;



            }
            catch (Exception ex)
            {
                throw new ApplicationException(ex.Message);

              //  return null;

            }
        }


        /// <summary>
        /// Executes a stored proc and returns a data set - usefull for when stored procs return multiple recordsets.
        /// </summary>
        /// <param name="strStoredProc"></param>
        /// <param name="_parameters"></param>
        /// <returns></returns>
        public DataSet ExecuteStoredProcDS(string strStoredProc, SqlParameter[] _parameters)
        {

            try
            {
                SqlCommand cm = new SqlCommand(strStoredProc, _conn);
                cm.CommandType = CommandType.StoredProcedure;
                // Check to see if there are any paramters being passed in.
                // If there are none, then do not add and execute the SP
                if ((((_parameters) != null)))
                {
                    foreach (SqlParameter p in _parameters)
                    {
                        cm.Parameters.Add(p);
                    }
                }
                SqlDataAdapter da = new SqlDataAdapter(cm);
                DataSet ds = new DataSet();

                da.Fill(ds);
                return ds;


            }
            catch (Exception ex)
            {
                throw new ApplicationException(ex.Message);
                // return null;
            }
        }

        public bool ExecuteStoredProc(string strStoredProc, SqlParameter[] _parameters)
        {
            try
            {
               
                    SqlCommand cm = new SqlCommand(strStoredProc, _conn);
                cm.CommandType = CommandType.StoredProcedure;

                if ((((_parameters) != null)))
                {
                    foreach (SqlParameter p in _parameters)
                    {
                        cm.Parameters.Add(p);

                    }
                }

                cm.ExecuteNonQuery();

                return true;


            }
            catch (Exception ex)
            {
                throw new ApplicationException(ex.Message);


            }
        }


        public bool GetRecordset(string strQuery)
        {
            SqlDataAdapter da;
            if (!(DataTbl == null))
            {
                DataTbl.Clear();
                DataTbl.Rows.Clear();
                DataTbl.Columns.Clear();
            }
            try
            {
                da = new SqlDataAdapter(strQuery, _conn);
                da.Fill(DataTbl);
                return true;
            }
            catch (SqlException ex)
            {
                MessageBox.Show("Error Getting Recordset: " + ex.Message); 
                return false;
            }
        }

        public System.Data.DataTable GetRecordsetDT(string strQuery)
        {
            SqlDataAdapter da;

            try
            {
                System.Data.DataTable dt = new System.Data.DataTable();
                da = new SqlDataAdapter(strQuery, _conn);
                da.Fill(dt);
                return dt;
            }
            catch (SqlException ex)
            {
                MessageBox.Show("Error Getting RecordsetDT: " + ex.Message); 
                return null;
            }
        }

        public DataTable GetDataTable(string strQuery)
        {
            try { 
                DataTable dataTable = new DataTable();

                SqlCommand cmd = _conn.CreateCommand();
                cmd.CommandText = strQuery;
                cmd.CommandType = CommandType.Text;

                // SqlDataReader need an open conncetion, so check and open it.
                if (_conn.State != ConnectionState.Open)
                    _conn.Open();

                // Read data by using Execute Reader
                SqlDataReader dr = cmd.ExecuteReader();//(CommandBehavior.CloseConnection);

                //Use data table load method to load data from data reader
                dataTable.Load(dr);

                return dataTable;
            }
            catch (SqlException ex)
            {
                MessageBox.Show("Error Getting RecordsetDT: " + ex.Message);
                return null;
            }
        }



        public void FillGrid(ref System.Windows.Forms.DataGridView grid, ref DataTable DataTable, params object[] strFieldList)
        {
            try
            {
                int x;
                int i;
                object[] objCells = new object[strFieldList.Length];
                grid.Rows.Clear();

                for (x = 0; x <= DataTable.Rows.Count - 1; x++)
                {
                    for (i = 0; (i <= strFieldList.Length-1); i++)
                    {
                        if ((strFieldList[i].ToString().Trim() != ""))
                        {
                            objCells[i] = Nvl(DataTable.Rows[x][strFieldList[i].ToString()],"");
                        }
                    }
                    grid.Rows.Add(objCells);
                }
            }
            catch (Exception ex)
            {
             //Currently not using FillGrid for anything yet.
                System.Windows.Forms.MessageBox.Show(ex.Message);
            }
        }

        public object Nvl(object vField, object vReplaceWith)
        {


            if (vField == System.DBNull.Value | (vField == null))
                return vReplaceWith;

            else
                return vField;



        }
        public bool ExecuteNonQuery(string strQuery)
        {
            SqlCommand objCommand;

            try
            {

                objCommand = new SqlCommand(strQuery, _conn);

               

                objCommand.ExecuteNonQuery();
                return true;
            }
            catch (SqlException ex)
            {
                MessageBox.Show("Error Executing Non-Query to SQL: " + ex.Message); 
                return false;
            }
            finally
            {
                objCommand = null;
            }
        }

        public bool ExecuteNonQuery(string strQuery, int iCommandTimeout)
        {
            SqlCommand objCommand;

            try
            {

                objCommand = new SqlCommand(strQuery, _conn);

                //Use an increased timeout
                objCommand.CommandTimeout = iCommandTimeout;

                objCommand.ExecuteNonQuery();
                return true;
            }
            catch (SqlException ex)
            {
                MessageBox.Show("Error Executing NonQuery to SQL: " + ex.Message); 
                return false;
            }
            finally
            {
                objCommand = null;
            }
        }


        public string BuildUpdateQry(string strTableName, params object[] objFieldsAndValues)
        {
            //Function returns a formatted insert query only. Makes NO connection to the database. 

            string qry;
            Int16 x;
            qry = "Update " + strTableName + " Set ";
            for (x = 0; x <= objFieldsAndValues.GetLength(0) - 1; x += 2)
            {

                if (objFieldsAndValues[x + 1] == null)
                    objFieldsAndValues[x + 1] = "";

                string str = objFieldsAndValues[x + 1].GetType().Name;
                str = str.ToUpper();
                switch (str)
                {
                    case "BOOLEAN":
                        if ((bool)objFieldsAndValues[x + 1] == true)
                            qry += objFieldsAndValues[x] + "= 1, ";
                        else
                            qry += objFieldsAndValues[x] + "= 0, ";
                        break;
                    case "INT32":
                    case "INT16":
                    case "INT64":
                        qry += objFieldsAndValues[x] + "=" + objFieldsAndValues[x + 1] + ", ";
                        break;
                    case "STRING":
                    case "DATETIME":
                        if (objFieldsAndValues[x + 1].ToString() == "NULL")
                            qry += objFieldsAndValues[x] + "= NULL, ";
                        else
                            qry += objFieldsAndValues[x] + "= '" + FixForApostrophe(objFieldsAndValues[x + 1].ToString()) + "', ";

                        break;
                    default:
                        qry += objFieldsAndValues[x] + "=" + objFieldsAndValues[x + 1] + ", ";
                        break;
                }

            }

            return qry.Substring(0, qry.Length - 2);
        }

        /// <summary>
        /// Note - FixForApostrophe is already built into this query generator
        /// </summary>
        /// <param name="strTableName"></param>
        /// <param name="objFieldsAndValues"></param>
        /// <returns></returns>
        public string BuildInsertQry(string strTableName, params object[] objFieldsAndValues)
        {
            // Function returns a formatted insert query only.  Makes NO connection to the database.

            Int16 x;
            string qry;
            qry = ("Insert Into " + (strTableName + " ("));
            for (x = 0; x <= objFieldsAndValues.GetLength(0) - 1; x += 2)
            {
                qry = (qry + (objFieldsAndValues[x] + ","));
            }
            qry = (qry.Substring(0, (qry.Length - 1)) + ") VALUES (");
            for (x = 1; x < objFieldsAndValues.GetLength(0); x += 2)
            {
                if ((objFieldsAndValues[x] == null))
                {
                    qry = (qry + (objFieldsAndValues[x] + "\'\',"));
                }
                else
                {
                    string str = objFieldsAndValues[x].GetType().Name;
                    str = str.ToUpper();
                    switch (str)
                    {
                        case "BOOLEAN":
                            if ((bool)objFieldsAndValues[x] == true)
                                qry = qry + "1,";
                            else
                                qry = qry + "0,";

                            break;
                        case "INT64":
                        case "INT32":
                        case "INT16":
                            qry = (qry + (objFieldsAndValues[x] + ","));
                            break;
                        case "STRING":
                            qry = (qry + ("\'" + (FixForApostrophe((string)objFieldsAndValues[x]) + "\',")));
                            break;
                        case "DATETIME":
                            qry = (qry + ("\'" + (objFieldsAndValues[x] + "\',")));
                            break;
                        default:
                            qry = (qry + (objFieldsAndValues[x] + ","));
                            break;
                    }
                }
            }
            return qry.Substring(0, (qry.Length - 1)) + ")";


        }


        public string FixForApostrophe(string strFieldContents)
        {
            //  replace each apostrophe with two apostrophes
            return strFieldContents.Replace("'", "''");

        }

        
        /// <summary>
        ///  Use ExecuteScalar method to return a single value from the database. 
        ///  The function returns the field object value
        /// </summary>
        /// <param name="strQuery"></param>
        /// <returns></returns>
        public object ExecuteScalar(string strQuery)
        {
            
            SqlCommand objCommand;
            try
            {
                

              
                objCommand = _conn.CreateCommand();
                objCommand.CommandText = strQuery;
                return objCommand.ExecuteScalar();
            }
            catch (SqlException  ex)
            {
             
                System.Windows.Forms.MessageBox.Show(ex.Message);
                return null;
            }
            finally
            {
                objCommand = null;
            }
        }

        /*

        /// <summary>
        /// Returns a SQLDataReader filled with one or more query results.  Separate multiple queries with semi-colons. 
        /// Use DataReader.NextResult to move between resultsets.
        /// You MUST close the reader when you're done with it.
        /// </summary>

        public SqlDataReader DataReader(string strQueries)
        {
            SqlCommand cmd;
            SqlConnection conn = new SqlConnection();
            conn.ConnectionString = _strConnectionString;
            conn.Open();
            cmd = new SqlCommand(strQueries, conn);
            return cmd.ExecuteReader();
        }

     
        

  


   

     

       
        */


    }

}